File: src/main/java/com/fengsheng/skill/JiZhi.java
package com.fengsheng.skill;

import com.fengsheng.Game;
import com.fengsheng.GameExecutor;
import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.phase.WaitForChengQing;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.concurrent.TimeUnit;

/**
 * 顾小梦技能【集智】：一名角色濒死时，或争夺阶段，你可以翻开此角色牌，然后摸四张牌。
 */
public class JiZhi extends AbstractSkill implements ActiveSkill {
    private static final Logger log = Logger.getLogger(JiZhi.class);

    @Override
    public SkillId getSkillId() {
        return SkillId.JI_ZHI;
    }

    @Override
    public void executeProtocol(Game g, Player r, GeneratedMessageV3 message) {
        if ((!(g.getFsm() instanceof FightPhaseIdle fsm) || r != fsm.whoseFightTurn)
                && (!(g.getFsm() instanceof WaitForChengQing fsm2) || r != fsm2.askWhom)) {
            log.error("现在不是发动[急智]的时机");
            return;
        }
        if (r.isRoleFaceUp()) {
            log.error("角色面朝上时不能发动[急智]");
            return;
        }
        var pb = (Role.skill_ji_zhi_tos) message;
        if ((r instanceof HumanPlayer humanPlayer) && !humanPlayer.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        r.incrSeq();
        r.addSkillUseCount(getSkillId());
        log.info(r + "发动了[急智]");
        for (Player p : g.getPlayers()) {
            if (p instanceof HumanPlayer player)
                player.send(Role.skill_ji_zhi_toc.newBuilder().setPlayerId(player.getAlternativeLocation(r.location())).build());
        }
        g.playerSetRoleFaceUp(r, true);
        r.draw(4);
        g.continueResolve();
    }

    public static boolean ai(FightPhaseIdle e, final ActiveSkill skill) {
        if (e.whoseFightTurn.isRoleFaceUp())
            return false;
        GameExecutor.post(e.whoseFightTurn.getGame(), () -> skill.executeProtocol(
                e.whoseFightTurn.getGame(), e.whoseFightTurn, Role.skill_ji_zhi_tos.getDefaultInstance()
        ), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/java/skill/JiZhi.kt
package com.fengsheng.skill

import com.fengsheng.Gameimport

com.fengsheng.GameExecutorimport com.fengsheng.HumanPlayerimport com.fengsheng.Playerimport com.fengsheng.phase.FightPhaseIdleimport com.fengsheng.phase.WaitForChengQingimport com.fengsheng.protos.Roleimport com.google.protobuf.GeneratedMessageV3import org.apache.log4j.Loggerimport java.util.concurrent.*
/**
 * 顾小梦技能【集智】：一名角色濒死时，或争夺阶段，你可以翻开此角色牌，然后摸四张牌。
 */
class JiZhi : AbstractSkill(), ActiveSkill {
    override fun getSkillId(): SkillId? {
        return SkillId.JI_ZHI
    }

    override fun executeProtocol(g: Game, r: Player, message: GeneratedMessageV3) {
        if ((g.fsm !is FightPhaseIdle || r !== fsm.whoseFightTurn)
            && (g.fsm !is WaitForChengQing || r !== fsm2.askWhom)
        ) {
            log.error("现在不是发动[急智]的时机")
            return
        }
        if (r.isRoleFaceUp) {
            log.error("角色面朝上时不能发动[急智]")
            return
        }
        val pb = message as Role.skill_ji_zhi_tos
        if (r is HumanPlayer && !r.checkSeq(pb.seq)) {
            log.error("操作太晚了, required Seq: " + r.seq + ", actual Seq: " + pb.seq)
            return
        }
        r.incrSeq()
        r.addSkillUseCount(skillId)
        log.info(r.toString() + "发动了[急智]")
        for (p in g.players) {
            (p as? HumanPlayer)?.send(
                Role.skill_ji_zhi_toc.newBuilder().setPlayerId(p.getAlternativeLocation(r.location())).build()
            )
        }
        g.playerSetRoleFaceUp(r, true)
        r.draw(4)
        if (g.fsm is FightPhaseIdle) fsm.whoseFightTurn = fsm.inFrontOfWhom
        g.continueResolve()
    }

    companion object {
        private val log = Logger.getLogger(JiZhi::class.java)
        fun ai(e: FightPhaseIdle, skill: ActiveSkill): Boolean {
            if (e.whoseFightTurn.isRoleFaceUp) return false
            GameExecutor.Companion.post(e.whoseFightTurn.game, Runnable {
                skill.executeProtocol(
                    e.whoseFightTurn.game, e.whoseFightTurn, Role.skill_ji_zhi_tos.getDefaultInstance()
                )
            }, 2, TimeUnit.SECONDS)
            return true
        }
    }
}